driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/OperationExecutorImpl.java

```diff
@@ -66,12 +66,13 @@
                         binding.release();
                         return Mono.error(new MongoClientException("Read preference in a transaction must be primary"));
                     } else {
-                        return Mono.<T>create(sink -> operation.executeAsync(binding, sinkToCallback(sink)))
-                                .doOnTerminate(binding::release)
-                                .doOnError((t) -> {
-                                    labelException(session, t);
-                                    unpinServerAddressOnTransientTransactionError(session, t);
-                                });
+                        return Mono.<T>create(sink -> operation.executeAsync(binding, (result, t) -> {
+                            binding.release();
+                            sinkToCallback(sink).onResult(result, t);
```

Comment:
Should we use try-finally in both places for releasing and calling `sinkToCallback`?

Reply:
like:
 
    try {
         sinkToCallback(sink).onResult(result, t);
    } finally {
        binding.release();
    }

If so, then no, because that would defeat the purpose of the change, which is to release the binding prior to the callback.  Maybe instead you mean:

    try {
         binding.release();
    } finally {
         sinkToCallback(sink).onResult(result, t);
    }

or 

    try {
         binding.release();
    } catch (RuntimeException e) }
         sinkToCallback(sink).onResult(null, e);    
    } finally {
         sinkToCallback(sink).onResult(result, t);
    }

?

Reply:
I mean `try {release} finally { sinkToCallback.onResult }`. I am sorry if my first comment appears to propose a different order, I did not think it is.

I am not opposed to the simplified approach because passing an exception from `release` to `onResult` is really verbose:
```
(result, t) -> {
    Throwable primaryProblem = t;//if t is not null, then this is the exception a user wants to see as the primary one, not the one from the release method
    try {
        binding.release();
    } catch (RuntimeException e) {
        if (primaryProblem == null) {
            primaryProblem = e;
        } else {
            primaryProblem.addSuppressed(e);
        }
    } finally {
        sinkToCallback(sink).onResult(result, primaryProblem);
    }
}
```

Whichever you chose is fine by me. We also have an alternative:
1.  specify that `release` must not throw anything but `IllegalStateException if the reference count is already 0`;
2. ensure that all the current implementations comply with 1;
3. not use try-finally with `release`.

Reply:
Actually, it's even worse:

1. On one hand, we actually should do
```java
finally {
    sinkToCallback(sink).onResult(primaryProblem == null ? result : null, primaryProblem);
}
```
2. On the other hand, doing 1 loses the result that was produced, which does not seem like a generally correct approach.

Given this, I am even more inclined to use the simplified version
```
try {
     binding.release();
} finally {
     sinkToCallback(sink).onResult(result, t);
}
```

Reply:
Done